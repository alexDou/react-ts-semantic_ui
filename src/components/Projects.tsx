import React, { Component, ReactElement } from 'react';
import { compose, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import {
    Segment, Container, Grid, Responsive, Header, Divider, Icon, Card, Image, Message
} from 'semantic-ui-react';
import { style } from 'typestyle';

import { AppState, ProjectsProps, StoreAction } from '@t/app';
import * as actions from '@store/actions';
import SearchContainer from '@containers/search';
import StatusReport from '@components/StatusReport';
import Footer from '@components/Footer';

const cardStyle = style({
    $nest: {
        '&.ui.card, &.ui.cards>.card': {
            width: '98%',
        },
    }
});

const segmentStyle = style({
    $nest: {
        '&.ui.segment': {
            marginBottom: '120px'
        }
    }
});

class Projects extends Component<AppState & ProjectsProps & RouteComponentProps, Readonly<keyof AppState>> {

    componentDidMount(): void {
        const { session, repos, getSearch } = this.props;
        const query = repos.query || '';

        if (!session.pending && repos.shouldFetch) {
            getSearch(query, 1);
        }
    }

    // shouldComponentUpdate(
    //     nextProps: Readonly<AppState & RouteComponentProps>,
    //     nextState: Readonly<AppState>
    // ): boolean {
    //
    //     return (
    //         nextProps.repos.query !== this.props.repos.query
    //         || nextProps.repos.page !== this.props.repos.page
    //         || nextProps.repos.per_page !== this.props.repos.per_page
    //         || nextProps.session.ok !== this.props.session.ok
    //     );
    // }

    componentDidUpdate(): void {
        const { session, repos, getSearch } = this.props;

        const pageNum = repos.page || 1;
        const query = repos.query || '';

        // get data to display from REST API
        if (!session.pending && repos.shouldFetch && !repos.display[pageNum] && query) {
            getSearch(query, pageNum);
        }
    }

    displayItems(): ReactElement {
        const { session, repos } = this.props;

        const pageNum = repos.page || 1;
        const items = repos.display[pageNum] || [];
        const notFound = repos.repos.total_count === 0;

        return (session.ok ?
            <Segment color={items.length ? 'green' : 'grey'}>
                {notFound ? <Message color="red" content="No repos found by this query" /> : '' }
                <Responsive>
                    <Grid celled="internally" stackable columns={3}>
                        {items.map((it: any): ReactElement =>
                            <Grid.Column key={it.id}>
                                <Card centered className={cardStyle}>
                                    <Card.Content>
                                        <Image floated="right" size="mini" src={it.owner.avatar_url} />
                                        <Card.Header as="a" href={it.owner.html_url}>
                                            {it.full_name.length < 25
                                                ? it.full_name
                                                : `${it.full_name.substr(0, 25)}...`}
                                        </Card.Header>
                                        <Card.Meta>{it.language}</Card.Meta>
                                        <Card.Description>
                                            Owner: <strong>{it.owner.login}</strong>
                                        </Card.Description>
                                        <Card.Description>
                                            Stars: <strong>{it.stargazers_count}</strong>
                                        </Card.Description>
                                    </Card.Content>
                                </Card>
                            </Grid.Column>
                        )}
                    </Grid>
                </Responsive>
            </Segment>
            : <></>
        )
    }

    handlePageChange = (page: number): void => {
        const { setPage } = this.props;

        setPage(page, true);
    };

    render(): ReactElement {
        const { session, repos } = this.props;

        const pageNum = repos.page || 1;

        return (
            <Segment className={segmentStyle}>
                <Divider horizontal>
                    <Header as="h4">
                        <Icon name="search" />
                        Search repositories
                    </Header>
                </Divider>
                <Responsive as={Container}>
                    <SearchContainer
                        size="small"
                        color="grey"
                        stacked={false}
                        disabled={false}
                        query={repos.query as string}
                    />
                </Responsive>

                <Divider horizontal>
                    <Header as="h4">
                        <Icon name="github square" />
                        Matching repositories
                    </Header>
                </Divider>
                <Responsive as={Container}>
                    {this.displayItems()}
                    <StatusReport { ...session } />
                </Responsive>
                <Footer
                    pages={repos!.repos!.total_count || 0}
                    active={pageNum}
                    pageChange={this.handlePageChange}
                    {...session}
                />
            </Segment>
        );
    }
}

const mapStateToProps = (state: AppState): AppState => state;

const mapDispatchToProps = (dispatch: Dispatch): Partial<AppState> => {
    return {
        getSearch: (query: string, page: number): StoreAction => dispatch(actions.repos.getSearch(query, page)),
        setPage: (page: number, shouldFetch: boolean): StoreAction => dispatch(actions.repos.setPage(page, shouldFetch)),
    };
};

export default compose(
    withRouter,
    connect(mapStateToProps, mapDispatchToProps)
)(Projects) as React.ComponentType;
