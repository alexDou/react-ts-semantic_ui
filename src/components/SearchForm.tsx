import React, { useState } from 'react';
import { Button, Form, Segment } from 'semantic-ui-react';
import { style } from 'typestyle';

import { SearchFormProps } from '@t/app';

const inputStyle = style({
    minWidth: '260px'
})

const SearchForm: React.FC<SearchFormProps> = (props) => {
    const { size, stacked, color, disabled, query, onFormSubmit } = props;
    const [q, setQ] = useState(query);

    const handleChange = (e: any) => setQ(e.target.value);
    const handleSubmit = () => onFormSubmit(q as string);

    const formStyle = style({
        $nest: {
            '&.ui.form': {
                maxWidth: size === 'small' ? '400px' : '100%',
                margin: size === 'small' ? '0 auto' : 'initial'
            }
        }
    });

    const form = () =>
        <>
            <Form.Input
                disabled={disabled}
                fluid
                icon="search"
                iconPosition="left"
                placeholder="discover best ideas"
                type="text"
                value={q}
                className={inputStyle}
                onChange={handleChange}
            />
            <Button disabled={disabled} basic color={color} fluid size={size}>
            Search
            </Button>
        </>

    return (
        <Form size={size} onSubmit={handleSubmit} className={formStyle}>
            <Segment stacked={!!stacked}>
                {size === 'small'
                    ? <Form.Group>{form()}</Form.Group>
                    : form()
                }
            </Segment>
        </Form>
    )
}

export default SearchForm;
