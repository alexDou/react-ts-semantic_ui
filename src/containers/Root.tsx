import React from 'react';
import { Provider } from 'react-redux';

import { RootProps } from '@t/app';
import store from '@store/store';

const Root = ({ iniStore, children }: RootProps) => {
    return (
        <Provider store={store(iniStore)}>
            {children}
        </Provider>
    );
};

export default Root;
