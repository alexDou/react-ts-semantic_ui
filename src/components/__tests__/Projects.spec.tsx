import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Root from '@containers/Root';
import Projects from '@components/Projects';
import SearchForm from '@components/SearchForm';
import SearchContainer from '@containers/search';
import {
    Segment, Card, Message
} from 'semantic-ui-react';

const activeState = {
    session: {
        pending: false,
        ok: true,
        failure: false,
        message: ''
    },
    repos: {
        display: {
            1: [{
                id: 'XZS-980',
                'full_name': 'test/repo',
                language: 'qq',
                'stargazers_count': 'infinity',
                owner: {
                    login: 'test',
                    'avatar_url': 'http://pp.dd'
                }
            }]
        },
        query: 'find something',
        page: 1,
        repos: [{
            "id": 123456,
            "node_id": "ABC=123",
            "name": "repo",
            "full_name": "author/repo",
            "private": false,
            "author": {
                "login": "author",
                "id": 654321,
                "node_id": "123=ABC",
                "avatar_url": "https://avatars0.githubusercontent.com/u/654321?v=4",
                "url": "https://api.github.com/users/author",
            },
            "html_url": "https://github.com/author/repo",
            "description": "Repo",
            "url": "https://api.github.com/repos/apollographql/react-apollo",
            "size": 12702,
            "stargazers_count": 5817,
            "watchers_count": 5817,
            "language": "TypeScript"
        }],
        shouldFetch: false
    }
}

describe('Projects component', () => {
    it('renders itself', () => {
        const proj = renderer.create(
            <Root iniStore={{} as any}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );
        const projInstance = proj.root;

        expect(projInstance.findByType(Segment)).toBeDefined();
        proj.unmount();
    });

    it('renders SearchContainer and SearchForm', () => {
        const proj = mount(
            <Root iniStore={{} as any}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );

        expect(proj.find(SearchContainer)).toHaveLength(1);
        expect(proj.find(SearchForm)).toHaveLength(1);
        proj.unmount();
    });

    it('has pending, failure & not found state', () => {
        const state = {
            session: {
                pending: true,
                ok: false,
                failure: false,
                message: ''
            },
            repos: {
                display: {},
                query: 'find something',
                repos: {
                    total_count: 100,
                    incomplete_results: false,
                    items: []
                },
                page: 1,
                shouldFetch: false
            }
        }

        const proj = mount(
            <Root iniStore={state}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );

        expect(proj.find(Message)).toHaveLength(1);
        proj.unmount();

        state.session.pending = false;
        state.session.ok = false;
        state.session.failure = true;
        state.session.message = '40x';

        const proj2 = mount(
            <Root iniStore={state}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );

        expect(proj2.find(Message)).toHaveLength(1);
        proj2.unmount();

        state.session.pending = false;
        state.session.ok = true;
        state.session.message = '';

        const proj3 = mount(
            <Root iniStore={state}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );

        expect(proj3.find(Card)).toHaveLength(0);
        expect(proj3.find(Message)).toHaveLength(0);
        proj3.unmount();
    });

    it('renders item from state', () => {
        const proj = mount(
            <Root iniStore={activeState}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );

        expect(proj.find(Card)).toHaveLength(1);
        proj.unmount();
    });

    it('matches snapshot', () => {
        const proj = renderer.create(
            <Root iniStore={activeState}>
                <BrowserRouter>
                    <Projects />
                </BrowserRouter>
            </Root>
        );

        expect(proj.toJSON()).toMatchSnapshot();
        proj.unmount();
    });
});
