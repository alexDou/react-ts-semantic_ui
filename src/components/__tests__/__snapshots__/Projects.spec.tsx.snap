// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects component matches snapshot 1`] = `
<div
  className="ui segment f4znuu9"
>
  <div
    className="ui horizontal divider"
  >
    <h4
      className="ui header"
    >
      <i
        aria-hidden="true"
        className="search icon"
        onClick={[Function]}
      />
      Search repositories
    </h4>
  </div>
  <div
    className="ui container"
  >
    <form
      className="ui small form f16e15ps"
      onSubmit={[Function]}
    >
      <div
        className="ui segment"
      >
        <div
          className="fields"
        >
          <div
            className="field fygxhk2"
          >
            <div
              className="ui fluid left icon input"
            >
              <input
                disabled={false}
                onChange={[Function]}
                placeholder="discover best ideas"
                type="text"
                value="find something"
              />
              <i
                aria-hidden="true"
                className="search icon"
                onClick={[Function]}
              />
            </div>
          </div>
          <button
            className="ui grey small basic fluid button"
            onClick={[Function]}
          >
            Search
          </button>
        </div>
      </div>
    </form>
  </div>
  <div
    className="ui horizontal divider"
  >
    <h4
      className="ui header"
    >
      <i
        aria-hidden="true"
        className="github square icon"
        onClick={[Function]}
      />
      Matching repositories
    </h4>
  </div>
  <div
    className="ui container"
  >
    <div
      className="ui green segment"
    >
      
      <div>
        <div
          className="ui stackable internally celled three column grid"
        >
          <div
            className="column"
          >
            <div
              className="ui centered card ftf0jd2"
              onClick={[Function]}
            >
              <div
                className="content"
              >
                <img
                  className="ui mini right floated image"
                  src="http://pp.dd"
                />
                <a
                  className="header"
                >
                  test/repo
                </a>
                <div
                  className="meta"
                >
                  qq
                </div>
                <div
                  className="description"
                >
                  Owner: 
                  <strong>
                    test
                  </strong>
                </div>
                <div
                  className="description"
                >
                  Stars: 
                  <strong>
                    infinity
                  </strong>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
